Sources/XcodeGenKit/PBXProjGenerator.swift

```diff
@@ -37,8 +37,8 @@ public class PBXProjGenerator {
         }
     }
 
-    func addObject(_ object: PBXObject) {
-        proj.objects.addObject(object)
+    func addObject(_ object: PBXObject, reference: String) {
```

Comment:
As we already have the instance of the object we can create the reference within this function using the dynamic type. Then we only have to pass in the id for the reference generator. This also lets us easily swap out the generator (as you are doing in this PR)

Reply:
In fact because a PBXObject doesn't have a reference anymore, we can just pass in the instance to the generator in xcproj instead of a class type. Then the class name could be looked up in the generator. You could make this change in https://github.com/xcodeswift/xcproj/pull/185.
Then later on the id param in the generator can be become optional, as we can create a hash from all the instance properties

Reply:
Good idea @yonaskolb 😛. It'll make easier to use hashes when we automate the conformance with Sourcery. Looking forward to it!

Reply:
@yonaskolb I came across this [part of](https://github.com/yonaskolb/XcodeGen/blob/11e1813474f0e18730a7c45328f79c57ece77f8c/Sources/XcodeGenKit/PBXProjGenerator.swift#L68) the generation where the targets references are generated in advance using the class type. Do you think I should also expose the class-based method or is there any other way you can think of to solve this issue?

Reply:
Oh, tricky! Yeah I guess both function would need to be there to support that. One with a `type` param and one with an `object` param.
It also means that if the `addObject` function creates the references, that the targets can't use that function and have to manually add the objects to the project.

BUT another way this could be solved is creating the actual targets instead of simple references and caching them instead, and then late modifying them. We can do that now as the objects are mutable, which they weren't when this was written